src/dns_parser.rs

```diff
@@ -2037,26 +2038,27 @@ impl DnsIncoming {
         v
     }
 
-    fn read_ipv4(&mut self) -> Ipv4Addr {
-        let bytes: [u8; 4] = (&self.data)[self.offset..self.offset + 4]
+    fn read_ipv4(&mut self) -> Result<Ipv4Addr> {
+        let bytes: [u8; 4] = self.data[self.offset..self.offset + 4]
```

Comment:
Since you are adding error handling, could you also check the possible out of range error?